### ISSUE TITLE
Avoid using CTEs in read/get requests

### ISSUE THREAD

**Issue by ruslantalpa** [PostgREST/postgrest#621]
@NikolayS pointed to an article that describes some problems CTEs generate (when the result is big)
http://adpgtech.blogspot.ru/2016/05/indiscriminate-use-of-ctes-considered.html?m=1

Currently we use a CTE like this

``` SQL
with pg_source as
    (select projects.id,
            row_to_json(clients.*) as client,
            coalesce(
                         (select array_to_json(array_agg(row_to_json(tasks)))
                          from
                              (select tasks.id
                               from tasks
                               where tasks.project_id = projects.id) tasks), '[]') as tasks
     from projects
     left outer join
         (select clients.id
          from clients) as clients on clients.id = projects.client_id
    )
select
    (select pg_catalog.count(1) from projects) as total_result_set,
    pg_catalog.count(t) as page_total,
    '' as header,
    coalesce(array_to_json(array_agg(row_to_json(t))), '[]')::character varying as body
from
    (select *
     from pg_source 
    ) t;

```

Not that the PR regarding limits is merged, it's possible to remove the CTE like this it like this

``` SQL
select
    (select pg_catalog.count(1) from projects) as total_result_set,
    pg_catalog.count(1) as page_total,
    '' as header,
    coalesce(array_to_json(array_agg(row_to_json(t))), '[]')::character varying as body
from
    (select projects.id,
                    row_to_json(clients.*) as client,
                    coalesce(
                                 (select array_to_json(array_agg(row_to_json(tasks)))
                                  from
                                      (select tasks.id
                                       from tasks
                                       where tasks.project_id = projects.id) tasks), '[]') as tasks
             from projects
             left outer join
                 (select clients.id
                  from clients) as clients on clients.id = projects.client_id

    ) t;

```

**Comment by ruslantalpa** [comment]
Almost all tests pass except 2 related to CSV.
I can not generate the CSV header directly in sql
```SQL
-- current query
with pg_source as (
    select 'a' as column_a, 'b' as column_b
)
select
    -- csv header
   (
    select coalesce(string_agg(a.k, ','), '')
    from
     (select json_object_keys(r)::text as k
      from
        (select row_to_json(hh) as r
         from pg_source as hh
         limit 1
        ) s
     ) a
  ) 
  || E'\n' ||
  --csv body
  coalesce(string_agg(substring(t::text, 2, length(t::text) - 2), ''), '') AS body
from (
    select * from pg_source
) t

-- new query
select
  -- csv header
   (
    select coalesce(string_agg(a.k, ','), '')
    from
     (select json_object_keys(r)::text as k
      from
        (select row_to_json(hh) as r
         from t as hh -- here pg_source changed into t but it does not work
         limit 1
        ) s
     ) a
  ) 
  || E'\n' ||
  --csv body
  coalesce(string_agg(substring(t::text, 2, length(t::text) - 2), ''), '') AS body
from (
  select 'a' as column_a, 'b' as column_b
) t
```
what the csv header section is trying to do is, get the first row only from source, turn it into json, get the keys and concatenate them
any ideas how to rewrite that section?

**Comment by ruslantalpa** [comment]
@begriffs 3.2.0 can be built without this feature/issue because the way we are using the cte here does not affect index usage, we need indexes working only inside the CTE, once we have the data from CTE ready, we only turn it into json, no other joins/filters are used.

**Comment by begriffs** [comment]
It will be nice to remove unnecessary CTEs. I'm forgetting: has that made it into the code yet or is there still that problem about generating CSV holding us back?

**Comment by ruslantalpa** [comment]
there is still the problem of CSV but as i said, in our particular case, there is no penalty (i think) for using the CTE so i am not supper stressed about it.
If someone can figure out the CSV thing, then it can be implemented

**Comment by bgamari** [comment]
I should mention that this issue hurts my use-case (a tool for analyzing GHC performance) quite badly with common queries which should be nearly instantaneous taking seconds.

**Comment by ruslantalpa** [comment]
What is the size of the total dataset, what is the size of the response, if you are doing embedding do you have indexes on foreign keys?
also why does a tool that analyzes ghc performance care how much a query takes, this is time spent in IO, should not affect the performance measurement of haskell code.

**Comment by ruslantalpa** [comment]
hop in gitter and provide details there please

**Comment by bgamari** [comment]
> What is the size of the total dataset, what is the size of the
> response, if you are doing embedding do you have indexes on foreign
> keys?
> 
The table being queried has a few tens of millions of rows. Indexes are
configured correctly. Several months ago I tracked the issue pretty
conclusively back to PostgreSQL being unable to optimize the query due
to the use of CTEs. I'm afraid I can't remember the details, but I
looked back in my `psql` history and found the following commands (which
I'm pretty certain arose out of this investigation),
```
$ psql ghc_perf
psql (9.6.1, server 9.5.4)
Type "help" for help.

ghc_perf=# select count(*) from results;
  count   
----------
 20749455
(1 row)

ghc_perf=# explain analyze SELECT x.column1 AS commit_sha
FROM (VALUES ('f53d761df9762232b54ec57a950d301011cd21f8', 'hello', 'asdf')) as x
WHERE NOT EXISTS (
    SELECT commits.commit_sha
    FROM results
    INNER JOIN commits ON (commits.commit_id = results.commit_id)
    INNER JOIN test_envs ON (test_envs.test_env_id = results.test_env_id)
    WHERE commits.commit_sha = x.column1
      AND test_envs.test_env_name = 'nomeata'
);
                                                                           QUERY PLAN                                                                            
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop Anti Join  (cost=0.98..70.54 rows=1 width=32) (actual time=0.030..0.030 rows=0 loops=1)
   ->  Result  (cost=0.00..0.01 rows=1 width=0) (actual time=0.001..0.001 rows=1 loops=1)
   ->  Nested Loop  (cost=0.98..70.50 rows=129 width=41) (actual time=0.028..0.028 rows=1 loops=1)
         ->  Nested Loop  (cost=0.41..9.48 rows=1 width=49) (actual time=0.017..0.017 rows=1 loops=1)
               ->  Index Scan using commits_commit_sha_idx on commits  (cost=0.41..8.43 rows=1 width=45) (actual time=0.012..0.012 rows=1 loops=1)
                     Index Cond: (commit_sha = ('f53d761df9762232b54ec57a950d301011cd21f8'::text))
               ->  Seq Scan on test_envs  (cost=0.00..1.04 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=1)
                     Filter: (test_env_name = 'nomeata'::text)
         ->  Index Only Scan using results_test_env_id_commit_id_idx1 on results  (cost=0.56..50.20 rows=1082 width=8) (actual time=0.010..0.010 rows=1 loops=1)
               Index Cond: ((test_env_id = test_envs.test_env_id) AND (commit_id = commits.commit_id))
               Heap Fetches: 0
 Planning time: 0.546 ms
 Execution time: 0.066 ms
(13 rows)


ghc_perf=# explain analyze WITH exclude_list AS (
    SELECT x.column1 AS commit_sha                                      
    FROM (VALUES ('f53d761df9762232b54ec57a950d301011cd21f8', 'hello')) AS x
)                                                                    
SELECT DISTINCT commits.commit_sha
FROM results
INNER JOIN commits ON (commits.commit_id = results.commit_id)                 
INNER JOIN test_envs ON (test_envs.test_env_id = results.test_env_id)
LEFT JOIN exclude_list ON exclude_list.commit_sha = commits.commit_sha
WHERE exclude_list.commit_sha IS NULL;
                                                                    QUERY PLAN                                                                    
--------------------------------------------------------------------------------------------------------------------------------------------------
 Unique  (cost=4796176.34..4895128.91 rows=51303 width=41) (actual time=54663.109..65431.037 rows=8200 loops=1)
   CTE exclude_list
     ->  Result  (cost=0.00..0.01 rows=1 width=0) (actual time=0.002..0.003 rows=1 loops=1)
   ->  Sort  (cost=4796176.33..4845652.61 rows=19790514 width=41) (actual time=54663.107..64318.378 rows=20745793 loops=1)
         Sort Key: commits.commit_sha
         Sort Method: external merge  Disk: 1034256kB
         ->  Hash Join  (cost=3263.10..1180146.68 rows=19790514 width=41) (actual time=84.522..13958.445 rows=20745793 loops=1)
               Hash Cond: (results.test_env_id = test_envs.test_env_id)
               ->  Hash Join  (cost=3262.03..908026.04 rows=19790514 width=45) (actual time=84.478..11364.620 rows=20745793 loops=1)
                     Hash Cond: (results.commit_id = commits.commit_id)
                     ->  Seq Scan on results  (cost=0.00..477574.00 rows=19790900 width=8) (actual time=0.006..7015.303 rows=20749455 loops=1)
                     ->  Hash  (cost=2169.75..2169.75 rows=51302 width=45) (actual time=83.890..83.890 rows=51074 loops=1)
                           Buckets: 65536  Batches: 2  Memory Usage: 2519kB
                           ->  Hash Anti Join  (cost=0.03..2169.75 rows=51302 width=45) (actual time=0.040..34.278 rows=51074 loops=1)
                                 Hash Cond: (commits.commit_sha = exclude_list.commit_sha)
                                 ->  Seq Scan on commits  (cost=0.00..1522.03 rows=51303 width=45) (actual time=0.005..12.466 rows=51075 loops=1)
                                 ->  Hash  (cost=0.02..0.02 rows=1 width=32) (actual time=0.012..0.012 rows=1 loops=1)
                                       Buckets: 1024  Batches: 1  Memory Usage: 9kB
                                       ->  CTE Scan on exclude_list  (cost=0.00..0.02 rows=1 width=32) (actual time=0.005..0.006 rows=1 loops=1)
               ->  Hash  (cost=1.03..1.03 rows=3 width=4) (actual time=0.015..0.015 rows=3 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 9kB
                     ->  Seq Scan on test_envs  (cost=0.00..1.03 rows=3 width=4) (actual time=0.005..0.007 rows=3 loops=1)
 Planning time: 1.363 ms
 Execution time: 65541.300 ms
(24 rows)

```

> also why does a tool that analyzes ghc performance care how much a
> query takes, this is time spent in IO, should not affect the
> performance measurement of haskell code.
>
It's a front-end for visualising long-term trends in GHC's
compile-time and runtime performance. The long query times make it
rather painful to use, unfortunately.

**Comment by begriffs** [comment]
Just noting here that in the chatroom we determined that there is actually some inefficiency in @bgamari's underlying schema. The CTE here is not the main culprit, although removing CTEs where possible is still a good idea.

**Comment by ruslantalpa** [comment]
i think the way we are using CTE now does not hurt the performance so i'll close this for now

**Comment by NikolayS** [comment]
The thing is related not only to performance. 

Yes, CTEs are optimization fences and in some cases lead to bad plans chosen, but what I'm more concerned about is that PostgREST now cannot be used with Citus tables, because Citus doesn't work with CTEs: https://docs.citusdata.com/en/v7.0/faq/faq.html?highlight=cte#are-there-any-postgresql-features-not-supported-by-citus

This might be the case for other sharding solutions as well.

So, this (decision to keep CTEs) excludes very good options to scale for big tables.

**Comment by ruslantalpa** [comment]
Agree, good reason to reopen and try removing the cte

**Comment by ruslantalpa** [comment]
@begriffs could you ask citus if they would look at our current query (we'll provide a schema and query) and suggest a new one that would work with citus?

**Comment by begriffs** [comment]
> PostgREST now cannot be used with Citus tables, because Citus doesn't work with CTEs

I wrote the FAQ entry you linked to actually, and most of the Citus docs. 😛 

Citus supports CTEs when it can determine that the query can be run in a single node of a database cluster. The problems happen when more complicated queries like CTEs must access data on multiple nodes.

Queries get run on a single Citus node when they filter a distributed table by its distribution column. This happens most commonly in a multi-tenant application when restricting the query to a single tenant. (See: [Distributed Data Modeling](https://docs.citusdata.com/en/v7.0/sharding/data_modeling.html))

Using part of that example:

```sql
CREATE TABLE accounts (
  id bigint,
  name text NOT NULL,
  image_url text NOT NULL,

  PRIMARY KEY (id)
);

CREATE TABLE ads (
  id bigint,
  account_id bigint,
  campaign_id bigint,
  name text NOT NULL,
  image_url text NOT NULL,
  target_url text NOT NULL,
  impressions_count bigint DEFAULT 0 NOT NULL,
  clicks_count bigint DEFAULT 0 NOT NULL,

  PRIMARY KEY (account_id, id),
  FOREIGN KEY (account_id) REFERENCES accounts
);

SELECT create_distributed_table('accounts', 'id');
SELECT create_distributed_table('ads',      'account_id');
```

Attempting to access `/ads` with PostgREST gives:

```json
{
  "hint":"Consider using an equality filter on the distributed table's partition column.",
  "details":null,
  "code":"0A000",
  "message":"could not run distributed query with complex table expressions"
}
```

However requesting `/ads?account_id=eq.1` is able to execute because `account_id` is the column by which ads was distributed in the cluster.

In short when a CTE filters by the distribution column in all its expressions then you're all good.

**Comment by steve-chavez** [comment]
https://www.postgresql.org/message-id/E1gv7F7-0003kE-Jb@gemulon.postgresql.org?fbclid=IwAR2JJckeKnEOMxMRyIKbeNO6jJ2LVzW2VX8AloKP9na6qiwIUOrgkt1YymM

**Comment by steve-chavez** [comment]
CTEs have been inlined since pg 12 so this is no longer an issue. It can also be proved with the new EXPLAIN support:

```bash
# On pg 11
# $ PGRST_DB_PLAN_ENABLED=true PGRST_DB_ANON_ROLE='postgrest_test_anonymous' postgrest-with postgresql-11 postgrest-run

$ curl 'localhost:3000/projects?select=*,clients(*)&id=eq.1' -H "Accept: application/vnd.pgrst.plan+text"

Aggregate  (cost=16.38..16.41 rows=1 width=112)
  CTE pgrst_source
    ->  Nested Loop Left Join  (cost=0.30..16.36 rows=1 width=72)
          ->  Index Scan using projects_pkey on projects  (cost=0.15..8.17 rows=1 width=40)
                Index Cond: (id = 1)
          ->  Index Scan using clients_pkey on clients  (cost=0.15..8.17 rows=1 width=36)
                Index Cond: (id = projects.client_id)
  ->  CTE Scan on pgrst_source  (cost=0.00..0.02 rows=1 width=72)

# on pg >= 12
# $ PGRST_DB_PLAN_ENABLED=true PGRST_DB_ANON_ROLE='postgrest_test_anonymous' postgrest-with postgresql-12 postgrest-run

$ curl 'localhost:3000/projects?select=*,clients(*)&id=eq.1' -H "Accept: application/vnd.pgrst.plan+text"

Aggregate  (cost=16.37..16.39 rows=1 width=112)
  ->  Nested Loop Left Join  (cost=0.30..16.35 rows=1 width=72)
        ->  Index Scan using projects_pkey on projects  (cost=0.15..8.17 rows=1 width=40)
              Index Cond: (id = 1)
        ->  Index Scan using clients_pkey on clients  (cost=0.15..8.17 rows=1 width=36)
              Index Cond: (id = projects.client_id)

# No CTE mention on the plan
```
